[DSE] Split memory intrinsics if they are dead in the middle

omern1 opened the pull request:
This patch enables Dead Store Elimination to split memory intrinsics that are dead in the middle into `Front` and `Rear`:

```
  // __Front__                 ___Rear___
  // | ------------- Dead ------------- |
  //         | --- Killing --- |
```

Resolves #72113

RKSimon commented:
Please can you reorder the commits so the new tests are added earlier so you can you show the codegen diff?

felipepiovezan commented:
The commit message should explain what is being changed here

felipepiovezan commented on llvm/test/DebugInfo/dse-split-memintrinsics.ll:
```diff
@@ -0,0 +1,305 @@
+; CHECK: [[DBG119]] = !DILocation(line: 6, column: 5, scope: [[DBG103]])
+; CHECK: [[DIASSIGNID120]] = distinct !DIAssignID()
+; CHECK: [[DBG121]] = !DILocation(line: 7, column: 5, scope: [[DBG103]])
+; CHECK: [[DBG122]] = !DILocation(line: 8, column: 1, scope: [[DBG103]])
```
are all of these actually necessary?

felipepiovezan commented on llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp:
```diff
@@ -554,6 +555,80 @@ static void shortenAssignment(Instruction *Inst, Value *OriginalDest,
+      ConstantInt::get(DeadWriteLength->getType(), FrontSize));
+  DeadIntrinsic->addDereferenceableParamAttr(0, FrontSize);
+
+  Value *Indices[1] = {ConstantInt::get(DeadWriteLength->getType(), RearStart)};
```
IIUC we have an array here for the conversion to ArrayRef later? There is an implicit conversion from `T` to `ArrayRef<T>`, so we probably don't need this

omern1 committed:
Pre-committed tests

omern1 committed:
[DSE] Split memory intrinsics that are dead in the middle

DSE can already shorten intrinsics which have dead fronts or rears.
This patch enables DSE to split memory intrinsics that
are dead in the middle into `Front` and `Rear`:

```
  // __Front__                 ___Rear___
  // | ------------- Dead ------------- |
  //         | --- Killing --- |
```

Resolves #72113

fhahn commented on llvm/test/DebugInfo/dse-split-memintrinsics.ll:
```diff
@@ -0,0 +1,305 @@
+;
+entry:
+  %blob = alloca [1000 x i8], align 16, !DIAssignID !112
+  call void @llvm.dbg.assign(metadata i1 undef, metadata !108, metadata !DIExpression(), metadata !112, metadata ptr %blob, metadata !DIExpression()), !dbg !113
```
please simplify the test, by dropping the debug info and unneeded lifetime s

omern1 replied:
I've simplified the test and removed the lifetimes, but I've not removed the debuginfo entirely because that specific test is for dbg.assign intrinsic updates.

fhahn commented on llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp:
```diff
@@ -554,6 +555,80 @@ static void shortenAssignment(Instruction *Inst, Value *OriginalDest,
+  // If Front and Rear are both bigger than the threshold they won't be inlined
+  // so this seems like a bad idea. If Dead is smaller than the threshold it
+  // will be inlined so this isn't a good idea.
+  if ((FrontSize > Threshold && RearSize > Threshold) || DeadSize < Threshold)
```
are all those combos covered by tests?

omern1 replied:
I've added tests to cover all scenarios, thanks.

omern1 committed:
Address review comments

- Remove unnecessary array
- Simplify debuginfo test
- Add tests to cover the profitability check
-

omern1 committed:
Clang format patch

bjope commented:
Thanks a lot for looking into this!

Not sure how much time I have to look at this before the end of the year. But I've downloaded the current patch to run some early testing.

A few things that I noticed already:
- The already existing helpers for trimming the start (tryToShorten) is involving some logic for atomic operations. I wonder if that is needed here as well (a bit unsure since the helper for trimming the end (tryToShortenEnd) isn't having any extra logic for atomics. So maybe it isn't needed, but perhaps something to look into unless you've figured out that it isn't needed.
- The heuristics for when to apply the transform definitely needs to be adjusted for my target (I don't think that getMaxMemIntrinsicInlineSizeThreshold() will be enough). We can for example easily memset/memcpy 1/2/4/8 bytes cheaply (a single store). While 7 bytes would require three stores. So even if for example the "frontsize" can be reduced to something that isn't a power of two, and smaller than eight, it would be profitable for our target to still memset/memcpy 8 bytes. And similarly for the "rearsize". So depending on the target it might be important to keep the start addresses of the two parts aligned, but sometimes also the number of bytes to write.
- I find the (DeadSize >= Threshold) requirement a bit restrictive (at least for us). One example would be if we have a memset writing X+Y+X bytes, with the Y bytes in the middle being dead. If for example X=8 and Y=1, then doing the split would result in two 8 byte stores, while a memset of 17 bytes would require 3 stores (in this example I assume that the target can do stores with align=1). So even if we normally would inline such a memset,  it would be profitable to split it up and skip writing the middle Y bytes as the front/rear would be cheaper as the sizes are well-aligned.

I understand that it could be hard to find a heuristic and TTI hooks that works well for any target (and specially for out-of-targets like mine). And I don't think that needs to be a goal with the first implementation either. Adding more hooks and conditions could be something that evolves over time, when tuning this for different targets. So some of the above comments are just to highlight some different scenarios that sprung to mind when considering  how to tune this for my target.

omern1 commented:
Hi @bjope,

Thanks for looking at this.

> The already existing helpers for trimming the start (tryToShorten) is involving some logic for atomic operations. I wonder if that is needed here as well (a bit unsure since the helper for trimming the end (tryToShortenEnd) isn't having any extra logic for atomics. So maybe it isn't needed, but perhaps something to look into unless you've figured out that it isn't needed.

I've looked at this now and I think the logic for trimming atomic operations is required here as well. I'll add it shortly.

>  find the (DeadSize >= Threshold) requirement a bit restrictive (at least for us). One example would be if we have a memset writing X+Y+X bytes, with the Y bytes in the middle being dead. If for example X=8 and Y=1, then doing the split would result in two 8 byte stores, while a memset of 17 bytes would require 3 stores (in this example I assume that the target can do stores with align=1). So even if we normally would inline such a memset, it would be profitable to split it up and skip writing the middle Y bytes as the front/rear would be cheaper as the sizes are well-aligned.

This is a good point, I'm inclined to add a check to let the transform run when `Front` and `Rear` are aligned.

omern1 committed:
Add check for atomic intrinsics

omern1 commented:
Apologies for the delay (I've been away), I've added the check for atomic intrinsics and an associated test like I mentioned earlier.

omern1 committed:
Empty commit to re-trigger CI

omern1 commented:
(Polite ping!)

omern1 committed:
Remove (DeadSize >= Threshold) requirement

omern1 commented:
Hi @bjope / @felipepiovezan / @fhahn, are there any other changes that you'd like me to make to this patch?

omern1 committed:
Update tests and add comment

omern1 commented:
I've now also removed the `DeadSize >= Threshold` condition.

bjope commented on llvm/test/Transforms/DeadStoreElimination/stores-of-existing-values.ll:
```diff
@@ -531,7 +533,9 @@ define void @test12_memset_other_store_in_between_partial_overlap(ptr %ptr) {
 ; CHECK-LABEL: @test12_memset_later_store_exceeds_memset(
-; CHECK-NEXT:    call void @llvm.memset.p0.i64(ptr align 1 [[PTR:%.*]], i8 0, i64 8, i1 false)
+; CHECK-NEXT:    [[TMP1:%.*]] = getelementptr inbounds i8, ptr [[PTR:%.*]], i64 5
+; CHECK-NEXT:    call void @llvm.memset.p0.i64(ptr align 1 dereferenceable(5) [[TMP1]], i8 0, i64 5, i1 false)
```
This looks wrong? We used to have a single memset writing 8 bytes, and now we got two memset writing 5+4=9 bytes.

omern1 replied:
Sorry for the delay, fixed in https://github.com/llvm/llvm-project/pull/75478/commits/bc32bcb6d8e7c10a8a141370540fda3adf932d30.

bjope commented on llvm/test/Transforms/MemCpyOpt/memcpy.ll:
```diff
@@ -246,7 +246,9 @@ define i32 @test5(i32 %x) nounwind ssp {
 ; CHECK-NEXT:    [[Y:%.*]] = alloca [[STRUCT_S:%.*]], align 16
-; CHECK-NEXT:    call void @llvm.memcpy.p0.p0.i64(ptr align 16 [[Y]], ptr align 16 @sS, i64 32, i1 false)
+; CHECK-NEXT:    [[TMP0:%.*]] = getelementptr inbounds i8, ptr [[Y]], i64 16
+; CHECK-NEXT:    call void @llvm.memcpy.p0.p0.i64(ptr align 16 dereferenceable(16) [[TMP0]], ptr align 16 @sS, i64 16, i1 false)
```
This also looks weird. Splitting up the memset in two, but we still write 16+16=32 bytes.

omern1 replied:
Fixed in https://github.com/llvm/llvm-project/pull/75478/commits/1805544b4472b4fb578c5494cbadbf56c1855333.

omern1 committed:
Prevent splitting if aligning causes all benefits to be lost

omern1 committed:
Make sure that other transforms operate on the correct intrinsic after splitting

omern1 committed:
Fix formatting

omern1 committed:
Merge remote-tracking branch 'upstream/main'

omern1 committed:
Update test to use debug records

omern1 committed:
Fix formatting

omern1 commented:
Ping! I think I've addressed all the review comments, are we happy to go ahead with this now?

omern1 commented:
Pinging again!

DianQK commented on llvm/test/Transforms/DeadStoreElimination/dead-middle-split.ll:
```diff
@@ -0,0 +1,94 @@
+;
+entry:
+  tail call void @llvm.memset.p0.i64(ptr noundef nonnull align 16 dereferenceable(1000) %X, i8 5, i64 1000, i1 false)
+  %arrayidx = getelementptr inbounds i8, ptr %X, i64 10
```
IIUC, this should be `%arrayidx = getelementptr inbounds i8, ptr %X, i64 16`, otherwise, this will be undefined behavior.

DianQK commented on llvm/test/DebugInfo/dse-split-memintrinsics.ll:
```diff
@@ -0,0 +1,113 @@
+!19 = !{i32 7, !"PIE Level", i32 2}
+!20 = !{i32 7, !"uwtable", i32 2}
+!21 = !{i32 7, !"debug-info-assignment-tracking", i1 true}
+!22 = !{!"clang version 18.0.0"}
```
Could you remove all the unnecessary metadata?

DianQK commented on llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp:
```diff
@@ -2054,7 +2147,7 @@ struct DSEState {
   }
 
-  bool removePartiallyOverlappedStores(InstOverlapIntervalsTy &IOL) {
+  bool removePartiallyOverlappedIntrinsicStores(InstOverlapIntervalsTy &IOL) {
```
I don't think we need to change this function name.

omern1 replied:
I'm not sure about that, I thought the name doesn't accurately convey what the function does, it works on store-like intrinsics not on `store` instructions but I'm happy to make the change if you feel strongly enough about it.

DianQK commented on llvm/test/Transforms/DeadStoreElimination/dead-middle-split.ll:
```diff
@@ -0,0 +1,94 @@
+; CHECK-SAME: ptr nocapture noundef writeonly [[X:%.*]]) {
+; CHECK-NEXT:  entry:
+; CHECK-NEXT:    [[TMP0:%.*]] = getelementptr inbounds i8, ptr [[X]], i64 976
+; CHECK-NEXT:    tail call void @llvm.memset.p0.i64(ptr noundef nonnull align 16 dereferenceable(24) [[TMP0]], i8 5, i64 24, i1 false)
```
We also need to use a smaller align value, can you add a test case for that?

DianQK commented on llvm/test/Transforms/DeadStoreElimination/dead-middle-split.ll:
```diff
@@ -0,0 +1,94 @@
+
+; RUN: opt < %s -passes=dse -S | FileCheck %s
+
+define void @overwrite_middle(ptr nocapture noundef writeonly %X) {
```
Do you need `nocapture noundef writeonly`?

omern1 committed:
Merge branch 'main' into main

omern1 committed:
Address review comments

DianQK commented on llvm/test/Transforms/DeadStoreElimination/dead-middle-split.ll:
```diff
@@ -0,0 +1,112 @@
+
+; RUN: opt < %s -passes=dse -S | FileCheck %s
+
+define void @overwrite_middle(ptr  %X) {
```
```suggestion
define void @overwrite_middle(ptr %X) {
```
Nits. Elsewhere, too.​

DianQK commented on llvm/test/Transforms/DeadStoreElimination/dead-middle-split.ll:
```diff
@@ -0,0 +1,112 @@
+; CHECK-NEXT:    ret void
+;
+entry:
+  call void @llvm.memset.element.unordered.atomic.p0.i64(ptr align 16 %X, i8 5, i64 1000, i32 4)
```
Can we split atomic?

dianqk reviewed (commented):
Have you handled the case where the alignments are different?​

For example,
```llvm
define void @overwrite_middle(ptr  %X) {
entry:
  tail call void @llvm.memset.p0.i64(ptr  align 16 dereferenceable(1000) %X, i8 5, i64 1000, i1 false)
  %arrayidx = getelementptr inbounds i8, ptr %X, i64 15
  tail call void @llvm.memset.p0.i64(ptr  align 1 dereferenceable(980) %arrayidx, i8 3, i64 980, i1 false)
  ret void
}
```

omern1 committed:
Address review comments

omern1 committed:
Update test

omern1 committed:
Merge remote-tracking branch 'upstream/main'

DianQK commented on llvm/test/Transforms/DeadStoreElimination/dead-middle-split.ll:
```diff
@@ -0,0 +1,112 @@
+; CHECK-SAME: ptr [[X:%.*]]) {
+; CHECK-NEXT:  entry:
+; CHECK-NEXT:    [[TMP0:%.*]] = getelementptr inbounds i8, ptr [[X]], i64 35
+; CHECK-NEXT:    tail call void @llvm.memset.p0.i64(ptr align 1 dereferenceable(15) [[TMP0]], i8 5, i64 15, i1 false)
+; CHECK-NEXT:    tail call void @llvm.memset.p0.i64(ptr align 1 dereferenceable(10) [[X]], i8 5, i64 10, i1 false)
```
Could you swap the two memset intrinsic? This will probably make the CPU cache happy?

DianQK commented on llvm/test/Transforms/DeadStoreElimination/dead-middle-split.ll:
```diff
@@ -0,0 +1,112 @@
+entry:
+  tail call void @llvm.memset.p0.i64(ptr  align 1 dereferenceable(50) %X, i8 5, i64 50, i1 false)
+  %arrayidx = getelementptr inbounds i8, ptr %X, i64 10
+  tail call void @llvm.memset.p0.i64(ptr  align 1 dereferenceable(45) %arrayidx, i8 3, i64 25, i1 false)
```
```suggestion
  tail call void @llvm.memset.p0.i64(ptr align 1 dereferenceable(45) %arrayidx, i8 3, i64 25, i1 false)
```

Maybe you can run `opt -S` to get a cleaner IR.

DianQK replied:
Please also remove the tail call.

dianqk reviewed (commented):
LGTM. Thanks! But I cannot review about atomic. :)
I mean that using a smaller align to reduce dead memory intrinsic: https://alive2.llvm.org/ce/z/LiGZUd. Maybe this could be another PR.

bjope commented:
I'm still a bit worried that this will result in lots of regressions, at least for my target.

I wonder if there perhaps should be some kind of TTI hook or a cmd line option to easily disable the whole feature (the latter could make it easier to benchmark by comparing results with/without the optimization enabled). I can of course add that downstream, so unless it will be used by in-tree targets it might seem weird to have such TTI hook.  But to make sure that this lands softly, maybe the feature should be disabled completely in the first version. Limiting any potential revert to flipping some big switch rather than having to revert all the code changes.

Things I'm concerned about when doing some test compiles with this enabled downstream are things like this:


The below would copy 29 bytes instead of 31. But for our target I think it will cost more to do two cross-addrspace memcpy compared to copy 2 more bytes in a single copy (both latency wise, and concerning the memory buses etc).

```
  call void @llvm.memcpy.p0.p2.i32(ptr %p, ptr addrspace(2) ..., i32 31, i1 false)
=>
  %p2 = getelementptr inbounds i8, ptr %p, i16 6
  call void @llvm.memcpy.p0.p2.i32(ptr %p2, ptr addrspace(2) ..., i32 25, i1 false)
  call void @llvm.memcpy.p0.p2.i32(ptr %p, ptr addrspace(2) ..., i32 4, i1 false)

```
The 144 byte memcpy in the example below is split into 123+20. So we only save a single byte. At least our target has optimized memcpy version copying several bytes per iterations in the inner loop (regardless of alignment). So splitting it up will just
result in two loops with an iteration count that isn't a multiple of 8 or 16 etc. This will be slower than just copying 18x8 bytes in a single memcpy.

```
  call void @llvm.memcpy.p0.p0.i32(ptr ..., ptr ..., i32 144, i1 false)
=>
  call void @llvm.memcpy.p0.p0.i32(ptr ..., ptr ..., i32 123, i1 false)
  call void @llvm.memcpy.p0.p0.i32(ptr ..., ptr ..., i32 20, i1 false)
```

For below example with memset being splitted we used to get three 8 byte stores. Now we get two 8 byte stores + one 4 byte store + one 1 byte store, plus extra pointer fiddeling, including increased register pressure. So even if we memset three bytes less the emitted asm looks much worse.
```
  call void @llvm.memset.p0.i16.i32(ptr noundef nonnull align 1 dereferenceable(24) %c, i16 0, i32 24, i1 false)
=>
   %rear2 = getelementptr inbounds i8, ptr %c, i16 4
   call void @llvm.memset.p0.i16.i32(ptr noundef nonnull align 1 dereferenceable(20) %rear2, i16 0, i32 20, i1 false)
   call void @llvm.memset.p0.i16.i32(ptr noundef nonnull align 1 dereferenceable(1) %ci, i16 0, i32 1, i1 false)
```

It is hard for a middle-end pass to know details about the target architecture, and in general we want it to canonicalize to something target independant. But I think it also might be hard to fuse thse memcpy:s back together in the backend. So a bit worried that even if on C-level the developer has decided to use a single memcpy (as a tuning) DSE now might mess up and split it in two memcpy even if it isn't feasible.

In the end I think we need more than the already existing getMaxMemIntrinsicInlineSizeThreshold (which we currently has set to UINT64_MAX for our target). So that might be another reason to land this "softly" by letting backend people add more tweaks to tune this and enable it in a more controlled way.

Maybe even this first implementation should include some more heuristics, such as not splitting an 144 byte memcpy into two "uneven" 123+20 byte pieces, just to save copying of a single byte? (The goal with https://github.com/llvm/llvm-project/issues/72113 was kind of to split up the memset/memcpy when we can avoid touching a larger chunk of memory twice.)

omern1 commented:
Thanks for the reviews and the explanation. Given the current situation I'm going to close this PR as I do not have the bandwith to work on this further at the moment. Hopefully someone else will be able to pick it up in the future.
